### PR TITLE
NED Refactoring

### PR DESCRIPTION
**Pull request by Mehrad0711**
This PR refactors and simplifies bootleg code. Also introduces the following changes:
- Adding Option to add qids to text as well useful for wikidata experiments.
- Updating `transform_dataset.py` to process and filter bootleg features for paraphrasing.
- Removing `*_plus_types` fields from examples and overwriting the main fields directly.
- Creating a class for Bootleg Annotator used in server/ predict mode to annotate examples on the fly

### CODE REVIEW

File: genienlp/arguments.py [hunk 1]
```diff
@@ -472,26 +480,29 @@ def parse_argv(parser):
...
         '--almond_domains', nargs='+', default=[], help='Domains used for almond dataset; e.g. music, books, ...'
     )
     parser.add_argument(
         '--ned_features',
         nargs='+',
         type=str,
-        default=[],
-        help='Features that will be extracted for each entity. Order is important',
+        default=['type_id', 'type_prob', 'qid'],
+        help='Features that will be extracted for each entity: "type" and "qid" are supported. Order is important',
```
Comment (gcampax) [inline genienlp/arguments.py:491]: We're changing the default here: what's the impact for current training configurations with and without Bootleg?

File: genienlp/arguments.py [hunk 2]
```diff
@@ -472,26 +480,29 @@ def parse_argv(parser):
...
     )
     parser.add_argument(
         '--ned_features_size',
         nargs='+',
         type=int,
-        default=[],
+        default=[1, 1, 1],
         help='Max length of each feature vector. All features are padded up to this length',
```
Comment (gcampax) [inline genienlp/arguments.py:498]: One part I still don't understand after all this time is the meaning of "feature vector". You have three features, and each feature is a sequence of words/classes coming out of Bootleg. What does "1" mean? What does it mean that the default value is 0 or 1.0?

File: genienlp/calibrate.py [hunk 3]
```diff
@@ -44,6 +44,58 @@
...
+
+    # Options to normalize scores for a given threshold
+    parser.add_argument(
+        '--threshold',
+        type=float,
+        default=None,
+        help='The threshold above (below) which scores are considered to be positive (negative).',
+    )
+    parser.add_argument('--precision', type=float, default=None, help='Set this if scores should be normalize by precision.')
+    parser.add_argument('--recall', type=float, default=None, help='Set this if scores should be normalize by recall.')
```
Comment (gcampax) [inline genienlp/calibrate.py:96]: Why did you move this around? Is there any functional change here? It's hard to tell.

File: genienlp/data_utils/bootleg.py [hunk 4]
```diff
@@ -33,108 +33,21 @@
...
-        model_name=args.bootleg_model,
-        verbose=False,
-    )
-    # collect all outputs now; we will filter later
-    bootleg_annotator.set_threshold(0.0)
-    setattr(bootleg_annotator, 'bootleg', bootleg)
-    return bootleg_annotator
-
-
 class Bootleg(object):
```
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:46]: So now we have "Bootleg", "BootlegAnnotator", and "BootlegAnnotator renamed to BootlegClass". This is beyond confusing. Can we pick better names, or add doc comments to distinguish?

File: genienlp/data_utils/bootleg.py [hunk 5]
```diff
@@ -33,108 +33,21 @@
...
 from bootleg.run import run_model
 from bootleg.utils.parser.parser_utils import parse_boot_and_emm_args
 
-from .database_utils import is_banned
+from .database_utils import is_banned, reverse_bisect_left
 
 logger = logging.getLogger(__name__)
 
 
-def reverse_bisect_left(a, x, lo=None, hi=None):
```
Comment (gcampax) [inline genienlp/data_utils/bootleg.py:46]: Why is this function not just called "binary_search"? What's with the comment about keeping it reverse-sorted? The function doesn't modify the input at all.

File: genienlp/paraphrase/scripts/transform_dataset.py [hunk 6]
```diff
@@ -129,6 +131,9 @@ def main(args):
     if args.remove_duplicates and args.no_duplication_columns is None:
         raise ValueError('You should specify columns that define duplication')
 
+    if args.bootleg_output_file:
+        bootleg_output_file = open(args.bootleg_output_file, 'w')
```
Comment (gcampax) [inline genienlp/paraphrase/scripts/transform_dataset.py:139]: Given you load the whole input file in memory anyway, you should load the input file before you open the output file. That way, the same file can be both input and output.

File: genienlp/predict.py [hunk 7]
```diff
@@ -512,10 +511,14 @@ def main(args):
         logger.info(f'Independent multi-GPU generation on following devices: {devices}')
         all_processes = []
         all_data_folders = split_folder_on_disk(args.data, len(devices))
+        if args.do_ned and args.ned_retrieve_method == 'bootleg':
+            all_bootleg_data_folders = split_folder_on_disk(args.bootleg_output_dir, len(devices))
```
Comment (gcampax) [inline genienlp/predict.py:515]: Why do we need this bootleg step? For train/dev, it's been run before, and for test, it will be run on the fly, right?

File: genienlp/run_bootleg.py [hunk 8]
```diff
@@ -83,11 +83,12 @@ def parse_argv(parser):
...
-        '--train_tasks', nargs='+', type=str, dest='train_task_names', help='tasks to use for training', required=True
-    )
-
-    parser.add_argument(
-        '--val_tasks', nargs='+', type=str, dest='val_task_names', help='tasks to collect evaluation metrics for'
+        '--tasks',
+        nargs='+',
+        type=str,
+        dest='train_task_names',
+        help='tasks to use for bootleg; we set dest to be train_task_names so we can resue previous code for arguments processsing',
```
Comment (gcampax) [inline genienlp/run_bootleg.py]: "we set dest to be train_task_names so we can resue previous code for arguments processsing" has nothing to do with the help for the command-line argument. Put it in a comment.

File: genienlp/arguments.py (hunk 2 above)
Comment (gcampax) [reply genienlp/arguments.py:498]: Also, what happens if the ned_ parameters given to `train` don't match those given to `run-bootleg`?

File: genienlp/tasks/almond_task.py [hunk 9]
```diff
@@ -455,16 +383,20 @@ def preprocess_field(self, sentence, field_name=None, answer=None, example_id=No
...
             ]
-        if 'type_prob' in self.args.ned_features and field_name != 'answer':
+        if self.args.do_ned and 'type_prob' in self.args.ned_features and field_name != 'answer':
             tokens_type_probs = [
                 [self.args.ned_features_default_val[1]] * self.args.ned_features_size[1] for _ in range(new_sentence_length)
             ]
+        if self.args.do_ned and 'qid' in self.args.ned_features and field_name != 'answer':
+            token_qids = [
+                [self.args.ned_features_default_val[2]] * self.args.ned_features_size[2] for _ in range(new_sentence_length)
+            ]
```
Comment (gcampax) [inline genienlp/tasks/almond_task.py:399]: I just noticed - why is this code in the Almond task and not in the general data loading code? How do we use NED for non-Almond tasks?

File: genienlp/util.py [hunk 10]
```diff
@@ -849,7 +853,7 @@ def load_config_json(args):
                 setattr(args, r, [0])
             elif r in ['ned_features', 'ned_features_size', 'ned_features_default_val', 'override_valid_metrics']:
                 setattr(args, r, [])
-            elif r == 'add_types_to_text':
+            elif r in ['add_types_to_text', 'add_qids_to_text']:
```
Comment (gcampax) [inline genienlp/util.py]: Use a tuple here, it's more idiomatic.

**Review by gcampax — COMMENTED** [review]
Overall, I like the new design and I think it's a good, long needed cleanup. I have some questions and some comments though.

**Review by s-jse — APPROVED** [review]
Looks good to me, after you address @gcampax's comments.

File: genienlp/calibrate.py (hunk 3 above)
Comment (Mehrad0711) [reply genienlp/calibrate.py:96]: No functional changes. I just moved the `parse_argv` to the top for predict, server, and calibrate to be consistent with the rest of the code.

Comment (gcampax) [reply genienlp/calibrate.py:96]: Next time, it will be helpful to have a dedicated commit for this change (with the magic words "No functional changes") so I won't worry about checking line by line.

Comment (Mehrad0711) [reply genienlp/calibrate.py:96]: I had this atomic commit:  "Move parse_argv method to the top consistent with the rest of the code" `10b9d9f0bb0252a8a51ccc026742b354b449785f` 
Good point on using magic words! Will do next time.

Comment (gcampax) [reply genienlp/calibrate.py:96]: Uh I swear I read through each commit. My bad, sorry!

Comment (Mehrad0711) [reply genienlp/calibrate.py:96]: No worries!

File: genienlp/arguments.py (hunk 1 above)
Comment (Mehrad0711) [reply genienlp/arguments.py:491]: For experiments that don't use ned (`do_ned` is False) there's no impact. 
For NED experiments, all features are processed by default. Now, If you want to use the features you would call `add_types_to_text` and/ or `add_qid_to_text`. I'm working on second cleanup for the code that we'll drop the `add_*_to_args` and add an argument to tell if we are adding features as text to the input or embedding ids. And then whatever features is passed as `ned_features` will be used automatically.

File: genienlp/arguments.py (hunk 2 above)
Comment (Mehrad0711) [reply genienlp/arguments.py:498]: Good catch. Since I moved out bootleg entirely out of almond task now, we don't need these arguments in `run_bootleg` anymore. I will remove them in this PR.

File: genienlp/data_utils/bootleg.py (hunk 4 above)
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:46]: Our `BootlegAnnotator` class is a wrapper for bootleg's `BootlegAnnotator` class. I can rename theirs to annotator in the import and add a docstring for our `Bootleg` and `BootlegAnnotator` classes.

Comment (gcampax) [reply genienlp/data_utils/bootleg.py:46]: What's the purpose of our BootlegAnnotator, and what's the purpose of the Bootleg class? Maybe we can find three distinct non-confusing names.

File: genienlp/tasks/almond_task.py (hunk 9 above)
Comment (Mehrad0711) [reply genienlp/tasks/almond_task.py:399]: I'm gradually moving all the code for NED out of almond task. Bootleg is now completely task-independent and can be used for other tasks too. The remaining code is for old approaches that use ES/ local database for NER. I will be moving these out also in the next PRs.

File: genienlp/paraphrase/scripts/transform_dataset.py (hunk 6 above)
Comment (Mehrad0711) [reply genienlp/paraphrase/scripts/transform_dataset.py:139]: Sure.

File: genienlp/predict.py (hunk 7 above)
Comment (Mehrad0711) [reply genienlp/predict.py:515]: This is needed for the paraphrase filtering step which is done on 4 gpus. The prepped bootleg files need to split the same way that input paraphrases are split.

File: genienlp/data_utils/bootleg.py (hunk 4 above)
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:46]: `Bootleg` class is a wrapper for all the functionalities we need from bootleg. It calls the appropriate functions from bootleg, processes the collected results, extract the needed features, and add those features to our examples.

`BootlegAnnotator` does the same but in a on-the-fly fashion useful for server mode when input data is unknown ahead of time.

I kinda like the current names but if the conflict is with the original `BootlegAnnotator` we can rename ours to `BootlegAnnotatorWrapper`.

File: genienlp/util.py (hunk 10 above)
Comment (Mehrad0711) [reply genienlp/util.py]: Fixed

File: genienlp/run_bootleg.py (hunk 8 above)
Comment (Mehrad0711) [reply genienlp/run_bootleg.py]: Fixed.

File: genienlp/data_utils/bootleg.py (hunk 5 above)
Comment (Mehrad0711) [reply genienlp/data_utils/bootleg.py:46]: This is binary_search for a reverse sorted list (i.e. descending order). I adopted the naming convention in `bisect` library. 
Yeah, the comment is misphrased. It should be: "Locate the insertion point for x in a to **maintain** its reverse sorted order"

**Comment by Mehrad0711** [comment]
I pushed a hotfix for bootleg to master and rebased this branch. That pushed all the previous commits ahead of your comments. The only new changes after your reviews are in the last two commits.

File: genienlp/arguments.py (hunk 2 above)
Comment (Mehrad0711) [reply genienlp/arguments.py:498]: The features are defined per token and include information about the wikidata qids, types, and probs assigned by bootleg . Each token can have up to 3 features defined by `ned_features`. Each of these features is a list of numbers. The size and default value for this list is defined by `ned_features_size` and `ned_features_default_val`.
For example, if size for type_id feature is 1, we only keep the first type associated with an entity Qid. 
If size for qid is 2, we keep the top 2 qids predicted by bootleg.
The default value is used to pad features up to the maximum size.
